### PR TITLE
test(inversion): gate 2,075 authored inversions on real bases, hermetically

### PR DESCRIPTION
**Pull request by nh13**
## What

A hermetic, PR-gating conformance suite for DNA inversions on real reference bases: 2,075 authored `inv` descriptions on `NM_004006.2` (the DMD coding sequence the recommendations use for their own worked inversion examples), swept as `c.{start}_{start+len-1}inv` for every `start` in `101, 108, …, 2999` and every `len` in `{4, 6, 8, 12, 16}`.

## Why

`DNA/inversion.md` is one of the few recommendation pages whose entire subject is sequence-dependent — the 3'rule (`:17`) reads the bases it would shift over, and the definition itself (`:5`) is a statement about a span and its reverse complement — and nothing in this repository was exercising it at scale against real bases. Every existing home declines for a structural reason:

- the manifest-backed conformance axes skip without `FERRO_MANIFEST`, which in CI is always;
- the harvested spec fixture normalizes through an empty `MockProvider`, so no sequence-dependent rule can fire;
- `examples/dump_normalized_corpus.rs` builds single-exon transcripts (#1478), so no generated row can cross an exon junction at all.

**125 of the 2,075 rows cross an exon junction** — precisely the geometry #1478 records as structurally unreachable in the generated corpora.

## Four layers, none of them redundant

1. **Exact outputs for all 2,075 rows**, in `tests/fixtures/inversion-sweep/cases.tsv`. This is the only layer that notices the 3'-shift *arithmetic* moving between two correct answers: `c.101_108inv -> c.102_107inv` silently becoming `-> c.103_106inv` keeps the same outcome class and the same census total — so the other layers can stay green while a normalized string a downstream consumer stores has moved.
2. **The sequence oracle**, and the layer that carries the weight. The output's own members, applied to the committed transcript bases, must produce the same sequence as replacing the authored span with its reverse complement. Neither side calls the normalizer, so it is not a self-consistency check, and it is indifferent to the *shape* of the answer — which is what lets it survive a repartitioning. Covers the 1,950 rows whose span lies in one exon; the junction-crossing 125 are excluded and counted, because a `c.` range spans its introns and their bases are not on the transcript axis.
3. **The shape invariant.** An authored `inv` normalizes to an `inv` (possibly 3'-shifted or narrowed), to a no-op `=` where the span equals its own reverse complement, or to allele members that are each on this transcript and each length-preserving — and to nothing else. The reverse complement of a span is exactly as long as the span, so a `del`/`dup`/`ins`/unbalanced-`delins` member asserts something `DNA/inversion.md:5` does not. Weaker than layer 2 but covers all 2,075 rows, and it is what makes a careless re-blessing of layer 1 still fail.
4. **The outcome census** — 1,434 unchanged / 460 shifted-still-inv / 26 palindromic no-op / 155 repartitioned, as named constants — so a bulk re-bless that shifts the *distribution* is visible in a diff and has to be argued for.

## What #1484 changed under this PR, and what was done about it

The invariant originally read "never split into allele members", which was true of the tree this corpus was authored against. #1484 widened the sequence-derived axis gate to `c.`/`n.`/`r.`, and **155 of the 2,075 rows now repartition**. The generator refused to pin them on the first run after the rebase — the refusal added in review, doing exactly its job — so this was found rather than absorbed.

`rulings[inversion-vs-two-delins-76-83]` records that the substitution case is decided and still splits ("#1230's substitution case is untouched and still splits", since `general.md:56` ranks substitution). So the shape rule was narrowed to what survives, and the falsifiable weight moved onto the new sequence oracle, which does not care what shape the answer takes as long as it denotes the same bases.

Of the 155: **87** split into substitutions only (the decided case), **43** into an `inv` core with flanking substitutions, and **25** into members that include a `delins` — the shape that same ruling says `inversion.md:5` governs. All 25 denote the correct bases, so it is a question about which legal spelling ships rather than about correctness. Filed as #1575 and pinned here as characterization, not endorsed.

## The pins are grounded, not merely self-consistent

`the_no_op_rows_are_exactly_the_self_reverse_complementary_spans` recomputes from the committed transcript bases which authored spans equal their own reverse complement, and asserts that set is *exactly* the set of rows that normalized to `=` — both directions, so a palindromic span left as an `inv` fails just as a non-palindromic span collapsed to a no-op does. That check answers to the spec's definition rather than to yesterday's output.

**Three rows are adjudications rather than characterization, and the module says which is which.** The spec publishes `NM_004006.2:c.5657_5660inv` and `NM_004006.2:c.4145_4160inv` outright (`DNA/inversion.md:27-34`, quoted verbatim in the test), and `c.5657_5660delinsTCAG` converges on the first — which `DNA/delins.md:5` settles by definitional exclusion ("**and which is not** a substitution or inversion"), consistent with the existing `inversion-vs-two-delins-76-83` ruling. The other 2,072 are characterization pins whose authority is the invariant plus the sequence grounding. `the_cited_clauses_are_still_verbatim` re-verifies all five quotes against the vendored checkout, so a submodule bump cannot leave them pointing at unrelated prose.

## Hermetic, and provably so

A `WindowProvider` over `tests/fixtures/inversion-sweep/reference-windows.json` serves the real bases with no `FERRO_MANIFEST`, so the gate cannot silently skip. Both window-fixture mitigations are in place and asserted, because the failure mode here is maximally deceptive — `src/normalize/mod.rs:3884-3897` turns a failed reference read into `Ok((input-unchanged, vec![]))`, and `Unchanged` is the *majority* outcome, so an under-serving slice would reproduce the pinned answer for 1,434 of 2,075 rows by accident:

- the generator's `serve_transcripts_whole` widens the transcript's window to the whole sequence, pinned hermetically by `the_committed_slice_serves_the_transcript_whole`;
- every pass runs through `AuditedProvider`, with the zero-failed-reads assertion evaluated **before** any verdict and a positive floor on served reads beside it.

**Deliberate-corruption check.** A single-base edit inside a palindromic span of the committed slice turns three gates red — the pins (`1 of 2075 pinned inversion outputs moved`), the census, and the sequence grounding — and turns the manifest-backed ground-truth tier red too. Separately, narrowing the transcript's window back to 4 bp (i.e. undoing `serve_transcripts_whole`) turns `the_committed_slice_serves_the_transcript_whole` red. One caveat worth stating: not *every* single-base flip is detected — a base at a position no row's trim decision depends on is inert, which is expected for a 13,993 bp transcript and is why the fidelity assertions exist alongside the pins rather than instead of them.

## One output that reads like a defect and is not

Four rows produce an intronic-offset endpoint:

```
c.1319_1334inv -> c.1322_1332-1inv
c.1326_1333inv -> c.1328_1332-1inv
c.1473_1484inv -> c.1475_1483-1inv
c.2789_2804inv -> c.2790_2804-1inv
```

A `c.` DNA description is genomic with intron offsets, so `c.2789_2804` denotes the genomic block from `c.2789` to `c.2804` **including intron 21** — 12,625 bp, not 16. Checked by hand against the captured window: that block is `[32472309, 32484933]` on `NC_000023.11`, its endpoints are `A`/`T` (complementary, so inverting leaves them untouched, hence the trim) and the next pair in is `C`/`T` (not complementary, so the trim stops after one base). The narrowing is symmetric on the genomic axis and identical in kind to the 481 within-exon narrowings. Recorded in the module docs so the next reader does not re-derive it.

## Verification

- `cargo fmt --all -- --check` — clean
- `cargo clippy --features dev --all-targets -- -D warnings` — exit 0, zero warnings
- `cargo nextest run --features dev -E 'binary(it)'` — `Summary [80.245s] 5514 tests run: 5514 passed (2 slow), 148 skipped`
- all ten new tests confirmed **run**, not skipped (named in the run output)
- `hermetic_normalization_matches_the_prepared_reference` run *with* `FERRO_MANIFEST`: all 2,075 rows normalize identically under the committed slice and the full prepared reference
- `extract_inversion_sweep_windows -- --check` against the manifest: both committed artifacts byte-identical to a fresh extraction

## Review fixes applied

Four findings from the CodeRabbit CLI lens. Two fixed, one rejected as mis-scoped, one verified as a **pre-existing repo-wide class** and deliberately left out of this PR.

**Fixed — the generator discarded its own fixed-point check** (flagged `major`). It called `normalize(&recorder, &rendered)?` to record the reads the gate needs, then dropped the result, leaving idempotency to `every_pinned_output_is_a_fixed_point` at test time. No hole — the invariant *was* guarded — but the failure landed after the artifact was written, so a careless regeneration commits the unstable row and the committed fixture then reads as though the instability were expected. That is exactly what the conformance refusal three lines below already refuses to do, and what `CLAUDE.md`'s "refuse, do not report" convention asks of a generator. The result is now compared and the generator bails. Control: inverting the condition refuses all 2,075 rows, so the check is reachable and not dead code.

**Fixed — the fixture provenance understated the window geometry** (flagged `minor`). It said `Margin: 64 bp`, which is true per range, but omitted that padded ranges within `MERGE_GAP = 128` are merged. A reader auditing the slice could reasonably assume every base in a window was read. It now states the padding, the merge gap, and the consequence: a window may span bases the pass never read, so its extent is not a claim about what was touched.

**Rejected — a second recorded window in `a_narrow_transcript_window_is_replaced_by_the_whole_sequence`** (flagged `trivial`). The test already pins the behaviour that matters (`serve_transcripts_whole` collapsing to the whole sequence); a second window exercises `merged_windows`, which has its own unit tests in `examples/common/recording.rs` including the `MERGE_GAP` boundary cases. Adding it here would test the shared helper through this generator rather than where it lives.

### Verified but deliberately not fixed here: `captured_from` is `prepared_at`, and that is a four-generator class

The lens asked for `reference_identity` instead of the timestamp. **The finding is right and I reproduced it**, but it is not a defect this PR introduces, and fixing it correctly is a separate change.

`prepared_at` differs between the two volumes that hold the *same* prepared reference:

| volume | `prepared_at` |
|---|---|
| scratch | `2026-06-17T09:44:18.428902+00:00` |
| backup | `2026-07-07T04:39:10.971467+00:00` |

Regenerating from the backup volume — which the layout notes explicitly endorse as a drop-in fallback when scratch is offline — reports the fixture out of date and produces a diff of **exactly two lines**, the timestamp, with every window byte and the whole census identical (`2075/2075 rows, 0 dropped; 5 windows, 204471 bp`). So `--check` says "the reference moved" when nothing moved.

The repo has in fact already ruled on this: `prepared_at` is **deliberately excluded** from the reference identity (#933) for precisely this reason, and #905 shipped per-artifact content stamps so an in-place change bumps the identity. `reference_identity` is a computed function in `src/prepare/identity.rs`, not a manifest field — so the finding's wording was off, but its substance holds.

Left out of this PR because `read_prepared_at` is the established pattern in **four** generators — this one plus `extract_split_member_separation_windows`, `extract_biocommons_windows` and `build_conformance_snapshot` — with a stated rationale in the comment. Changing one leaves it inconsistent with three siblings and each fix regenerates its own committed fixture. That is a coherent follow-up, not a rider on an inversion gate.

## Notes for review

- The generator is registered in `Cargo.toml` with `test = true`, since it carries `#[cfg(test)]` units guarding `serve_transcripts_whole`.
- `tests/it/main.rs` gains its `mod inversion_sweep;` line — without it the file is never compiled and CI stays green regardless.
- Fixture sizes: `cases.tsv` 107 KB, `reference-windows.json` 227 KB (one transcript served whole, 13,993 bp, plus four padded `NC_000023.11` windows totalling ~190 kb that the pass genuinely reads). Both under the 500 KB pre-commit ceiling.

Representation-Change: none. Test-only — nothing under `src/normalize/`, `src/hgvs/`, `src/spdi/` or `src/project/` is touched (verified against the staged diff); the only `src/` addition is `src/conformance/inversion_sweep.rs`, a fixture schema and outcome classifier with no runtime callers. No normalized output moves.

## Summary by CodeRabbit

* **New Features**
  * Added comprehensive inversion-sweep conformance coverage for DMD variants, including normalization, reverse-complement handling, and outcome classification.
  * Added tooling to generate, validate, and check inversion-sweep cases and reference windows from a recording manifest.
* **Tests**
  * Added integration coverage for 2,075 inversion variants, specification examples, fixed-point behavior, transcript access, and reference agreement.
  * Added validation for malformed fixtures, expected outcomes, and deterministic artifact consistency.
* **Documentation**
  * Documented generated test fixtures, regeneration commands, and artifact drift checks.

### CODE REVIEW

**Comment by nh13** [comment]
@coderabbitai pause

**Comment by nh13** [comment]
@coderabbitai review

File: tests/it/inversion_sweep.rs [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+            "hermetic_normalization_matches_the_prepared_reference: skipping — no FERRO_MANIFEST. \
+             The per-PR gates are every_authored_inversion_produces_its_pinned_output and \
+             no_authored_inversion_leaves_the_inversion_family, which never skip."
+        );
+        return;
+    };
+    if !manifest.exists() {
+        eprintln!("hermetic_normalization_matches_the_prepared_reference: skipping — FERRO_MANIFEST does not exist");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (nh13) [reply tests/it/inversion_sweep.rs:977]: Not applied: `eprintln!` + `return` **is** this suite's established mechanism for a manifest-absent tier — `biocommons_normalize_tests.rs`, `hgvs_rs_projection_tests.rs`, `issue_1086_cross_axis_projection.rs`, `issue_1086_c_axis_genomic_context.rs`, `issue_1217_mito_terminus_insertion_clamp.rs` and `issue_1539_split_member_separation.rs` all do exactly this. There is no skip helper to reach for, and nextest has no runtime skip; making this one file the outlier would be a divergence, not a fix.

The `tests/**` path instruction asks that a skip path not "return silently without printing why" — this one prints why, and names the two gates that never skip (`every_authored_inversion_produces_its_pinned_output`, `no_authored_inversion_leaves_the_inversion_family`) so a reader knows what is still enforcing. Introducing a real skip mechanism is worth doing across all six suites at once, not in this PR.

**Comment by nh13** [comment]
@coderabbitai review
